File: termpipe_mcp/tools/process.py
"""
Process management tools for TermPipe MCP Server.
"""

import os
import subprocess
from datetime import datetime
from typing import Optional, Dict, Any


class ProcessManager:
    """Track running processes for interactive sessions (REPLs, background jobs)"""
    
    def __init__(self):
        self.processes: Dict[int, Dict[str, Any]] = {}
    
    def add(self, pid: int, proc: subprocess.Popen, command: str):
        self.processes[pid] = {
            "command": command,
            "process": proc,
            "started": datetime.now(),
            "blocked": True
        }
    
    def get(self, pid: int) -> Optional[Dict[str, Any]]:
        return self.processes.get(pid)
    
    def is_blocked(self, pid: int) -> bool:
        proc_info = self.get(pid)
        return proc_info.get("blocked", False) if proc_info else False
    
    def set_blocked(self, pid: int, blocked: bool):
        if pid in self.processes:
            self.processes[pid]["blocked"] = blocked
    
    def remove(self, pid: int):
        if pid in self.processes:
            del self.processes[pid]
    
    def list_all(self) -> list[dict]:
        result = []
        for pid, info in self.processes.items():
            proc = info["process"]
            result.append({
                "pid": pid,
                "command": info["command"],
                "started": info["started"].isoformat(),
                "running": proc.poll() is None,
                "blocked": info.get("blocked", False),
                "return_code": proc.returncode
            })
        return result
    
    def cleanup_finished(self):
        finished = [pid for pid, info in self.processes.items() 
                   if info["process"].poll() is not None]
        for pid in finished:
            self.remove(pid)


# Global instance
process_manager = ProcessManager()


def register_tools(mcp):
    """Register process management tools with the MCP server."""
    
    @mcp.tool()
    def list_sessions() -> str:
        """
        List all active terminal sessions (running processes).
        Shows session status including PID, blocked status, and runtime.
        """
        process_manager.cleanup_finished()
        sessions = process_manager.list_all()
        
        if not sessions:
            return "📭 No active sessions"
        
        output = f"🔵 Active Sessions ({len(sessions)}):\n\n"
        for s in sessions:
            status = "⏸️  Blocked" if s["blocked"] else "▶️  Running"
            output += f"  PID {s['pid']}: {s['command'][:40]}\n"
            output += f"       Status: {status} | Code: {s['return_code']}\n\n"
        
        return output

    @mcp.tool()
    def read_process_output(pid: int) -> str:
        """
        Read output from a running process (REPL/interactive session).
        """
        proc_info = process_manager.get(pid)
        if not proc_info:
            return f"[Error: No session found with PID {pid}]"
        
        proc = proc_info["process"]
        poll_result = proc.poll()
        if poll_result is not None:
            process_manager.remove(pid)
            stdout = proc.stdout.read() if proc.stdout else ""
            return f"✅ Process finished (exit code: {poll_result})\n\n{stdout}"
        
        try:
            import fcntl
            stdout = proc.stdout
            if stdout:
                flags = fcntl.fcntl(stdout.fileno(), fcntl.F_GETFL)
                fcntl.fcntl(stdout.fileno(), fcntl.F_SETFL, flags | os.O_NONBLOCK)
                
                try:
                    output = stdout.read()
                    if output:
                        process_manager.set_blocked(pid, False)
                        return f"📤 Output:\n{output.decode() if isinstance(output, bytes) else output}"
                except BlockingIOError:
                    pass
                
                process_manager.set_blocked(pid, True)
                return f"⏳ Process {pid} is waiting for input (REPL prompt detected)"
            
            return f"⏳ Process {pid} running (no stdout available)"
            
        except Exception as e:
            return f"[Error reading output: {str(e)}]"

    @mcp.tool()
    def interact_with_process(pid: int, input: str) -> str:
        """
        Send input to a running process (REPL, interactive shell, etc).
        """
        proc_info = process_manager.get(pid)
        if not proc_info:
            return f"[Error: No session found with PID {pid}]"
        
        proc = proc_info["process"]
        
        if proc.poll() is not None:
            process_manager.remove(pid)
            return f"[Error: Process {pid} is no longer running]"
        
        try:
            if proc.stdin:
                proc.stdin.write(input)
                if not input.endswith("\n"):
                    proc.stdin.write("\n")
                proc.stdin.flush()
            
            process_manager.set_blocked(pid, True)
            return f"📤 Sent to PID {pid}:\n{input}\n\n💡 Use read_process_output({pid}) to see response"
            
        except BrokenPipeError:
            process_manager.remove(pid)
            return f"[Error: Process {pid} stdin closed]"
        except Exception as e:
            return f"[Error: {str(e)}]"

    @mcp.tool()
    def force_terminate(pid: int) -> str:
        """Force terminate a running process by PID."""
        proc_info = process_manager.get(pid)
        if not proc_info:
            try:
                os.kill(pid, 9)
                return f"✅ Killed process {pid} (was not tracked)"
            except ProcessLookupError:
                return f"[Error: No process found with PID {pid}]"
            except Exception as e:
                return f"[Error: {str(e)}]"
        
        try:
            proc = proc_info["process"]
            proc.kill()
            process_manager.remove(pid)
            return f"✅ Terminated process {pid}"
        except Exception as e:
            return f"[Error: {str(e)}]"

    @mcp.tool()
    def list_processes() -> str:
        """List all running processes (system-wide)."""
        try:
            proc = subprocess.run(
                ["ps", "aux", "--sort=-%mem"],
                capture_output=True,
                text=True,
                timeout=10
            )
            
            lines = proc.stdout.strip().split("\n")
            output = "🖥️  System Processes (sorted by memory):\n\n"
            output += "\n".join(lines[:51])
            
            if len(lines) > 51:
                output += f"\n... and {len(lines) - 51} more (use 'ps aux' for full list)"
            
            return output
            
        except Exception as e:
            return f"[Error listing processes: {str(e)}]"

File: go-mcp/pkg/tools/process/process.go
package process

import (
	"context"
	"fmt"
	"os/exec"
	"strings"
	"sync"
	"syscall"
	"time"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

type ProcInfo struct {
	Command   string
	Process   *exec.Cmd
	Started   time.Time
	Blocked   bool
}

var (
	processes = make(map[int]*ProcInfo)
	procMu    sync.Mutex
)

func RegisterTools(s *server.MCPServer) {
	// list_sessions
	listTool := mcp.NewTool("list_sessions",
		mcp.WithDescription("List all active terminal sessions (running processes)."),
	)
	s.AddTool(listTool, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		procMu.Lock()
		defer procMu.Unlock()

		if len(processes) == 0 {
			return mcp.NewToolResultText("📭 No active sessions"), nil
		}

		var output strings.Builder
		output.WriteString(fmt.Sprintf("🔵 Active Sessions (%d):\n\n", len(processes)))
		for pid, info := range processes {
			status := "▶️  Running"
			if info.Blocked {
				status = "⏸️  Blocked"
			}
			output.WriteString(fmt.Sprintf("  PID %d: %s\n", pid, info.Command))
			output.WriteString(fmt.Sprintf("       Status: %s\n\n", status))
		}
		return mcp.NewToolResultText(output.String()), nil
	})

	// read_process_output
	readOutputTool := mcp.NewTool("read_process_output",
		mcp.WithDescription("Read output from a running process."),
		mcp.WithNumber("pid", mcp.Required()),
	)
	s.AddTool(readOutputTool, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		pid := request.GetInt("pid", 0)
		return mcp.NewToolResultText(fmt.Sprintf("[Error: read_process_output not fully implemented for Go port yet; PID %d]", pid)), nil
	})

	// interact_with_process
	interactTool := mcp.NewTool("interact_with_process",
		mcp.WithDescription("Send input to a running process."),
		mcp.WithNumber("pid", mcp.Required()),
		mcp.WithString("input", mcp.Required()),
	)
	s.AddTool(interactTool, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		pid := request.GetInt("pid", 0)
		return mcp.NewToolResultText(fmt.Sprintf("[Error: interact_with_process not fully implemented for Go port yet; PID %d]", pid)), nil
	})

	// force_terminate
	terminateTool := mcp.NewTool("force_terminate",
		mcp.WithDescription("Force terminate a running process by PID."),
		mcp.WithNumber("pid", mcp.Required()),
	)
	s.AddTool(terminateTool, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		pid := request.GetInt("pid", 0)
		
		procMu.Lock()
		if info, exists := processes[pid]; exists {
			if info.Process != nil && info.Process.Process != nil {
				info.Process.Process.Kill()
			}
			delete(processes, pid)
		}
		procMu.Unlock()
		
		syscall.Kill(pid, syscall.SIGKILL)
		return mcp.NewToolResultText(fmt.Sprintf("✅ Terminated process %d", pid)), nil
	})

	// list_processes
	listProcsTool := mcp.NewTool("list_processes",
		mcp.WithDescription("List all running processes (system-wide)."),
	)
	s.AddTool(listProcsTool, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		cmd := exec.Command("ps", "aux", "--sort=-%mem")
		out, err := cmd.CombinedOutput()
		if err != nil {
			return mcp.NewToolResultText(fmt.Sprintf("[Error: %v]", err)), nil
		}
		lines := strings.Split(strings.TrimSpace(string(out)), "\n")
		var output strings.Builder
		output.WriteString("🖥️  System Processes (sorted by memory):\n\n")
		
		end := 51
		if len(lines) < 51 { end = len(lines) }
		for i := 0; i < end; i++ {
			output.WriteString(lines[i] + "\n")
		}
		if len(lines) > 51 {
			output.WriteString(fmt.Sprintf("\n... and %d more (use 'ps aux' for full list)", len(lines)-51))
		}
		return mcp.NewToolResultText(output.String()), nil
	})
}
